### ISSUE TITLE
[train][bug] `normalize_advantages_dict` computes mean over padding positions

### ISSUE THREAD

**Issue by CharlieFRuan** [NovaSky-AI/SkyRL#1305]
## Summary

`normalize_advantages_dict` in `ppo_utils.py` uses an unmasked `advantages.mean()` to compute the global mean, but uses a masked computation for the variance. Padding positions (where `response_mask == 0`) contaminate the mean, producing incorrect normalized advantages.

## Location

[`skyrl/backends/skyrl_train/utils/ppo_utils.py:127-144`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L127-L144)

```python
@torch.no_grad()
def normalize_advantages_dict(data: TrainingInputBatch) -> TrainingInputBatch:
    advantages: Float[torch.Tensor, "batch_size seqlen"] = data["advantages"]
    response_masks: Float[torch.Tensor, "batch_size seqlen"] = data["response_mask"]
    num_actions: float = response_masks.sum()
    # mean
    mean: float = advantages.mean()                                    # BUG: includes padding
    # std
    std: float = ((advantages - mean).pow(2) * response_masks).sum()   # correct: masked
    rstd: float = (std / num_actions).clamp(min=1e-8).rsqrt()

    data["advantages"] = (advantages - mean) * rstd
    return data
```

Line 138 computes the mean over **all** tensor elements. Line 140 correctly masks with `response_masks`. The mean and variance use inconsistent masking.

## Call path

1. **Training loop** ([`trainer.py:271`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/trainer.py#L271)): `self.compute_advantages_and_returns(training_input)`
2. **Dispatch** ([`ppo_utils.py:1190-1214`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1190-L1214)): `compute_advantages_and_returns()` calls the registered advantage estimator (GAE, GRPO, RLOO, or REINFORCE++)
3. **Normalization** ([`trainer.py:277-278`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/trainer.py#L277-L278)): if `cfg.trainer.algorithm.advantage_batch_normalize` is enabled:
   ```python
   training_input = normalize_advantages_dict(training_input)
   ```

The same path exists in [`fully_async_trainer.py:515-516`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/train/fully_async_trainer.py#L515-L516).

## Severity by estimator

The impact depends on what the advantage estimator leaves in padding positions:

| Estimator | Padding positions | Impact |
|---|---|---|
| **GAE** ([`ppo_utils.py:1128`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1128)) | **Non-zero garbage** — [`masked_whiten`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L165-L171) normalizes using masked stats but does not zero out padding positions | **High** — mean is contaminated by arbitrary values from the GAE recurrence |
| **GRPO** ([`ppo_utils.py:1096`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1096)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — zeros dilute the mean toward zero proportional to padding ratio |
| **RLOO** ([`ppo_utils.py:1096`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1096)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — same dilution effect |
| **REINFORCE++** ([`ppo_utils.py:1043`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L1043)) | Zero — explicitly multiplies by `response_mask` | **Low-moderate** — same dilution effect |

For the zero-padding estimators, including padding zeros still biases the mean: if the true masked mean is positive, the diluted mean is closer to zero, so the subtracted mean is too small. The bias scales with `padding_tokens / total_tokens`.

## Suggested fix

Use the already-available [`masked_mean`](https://github.com/NovaSky-AI/SkyRL/blob/ad9105b15f959abffc9c16e07e26b4cbc2274f6b/skyrl/backends/skyrl_train/utils/ppo_utils.py#L33) (imported from `torch_utils` at line 33):

```python
mean: float = masked_mean(advantages, response_masks)
```

This makes the mean computation consistent with the variance computation, averaging only over valid response positions.

**Comment by tamoghnokandar** [comment]
I can take up this issue.

**Comment by discobot** [comment]
Note the code has moved since this was filed: `normalize_advantages_dict` is now `RayPPOTrainer._normalize_advantages` on main (inherited by `FullyAsyncRayPPOTrainer`), bug intact — fix with a padding-invariance regression test in #1782.
